src/dataflow/src/sink/tail.rs

```diff
@@ -71,19 +79,15 @@ pub fn tail<G>(
             results.sort_by_key(|(time, _)| *time);
             let mut results: Vec<Row> = results.into_iter().map(|(_, row)| row).collect();
 
-            if connector.emit_progress {
-                if let Some(upper) = batch_upper(batches.last()) {
-                    // The user has requested progress messages and there's at least one
-                    // batch. All of the batches might have zero rows, so we do not depend on
-                    // results at all. Another benefit of using upper (instead of the largest row
-                    // time) is that the batch's upper may be larger than the row time.
-                    packer.push(Datum::Decimal(Significand::new(i128::from(upper))));
-                    packer.push(Datum::True);
-                    // Fill in the diff column and all table columns with NULL.
-                    for _ in 0..(connector.object_columns + 1) {
-                        packer.push(Datum::Null);
-                    }
-                    results.push(packer.finish_and_reuse());
+            if let Some(batch) = batches.last() {
+                let progress_row = update_progress(
```

Comment:
This ensures that there will be a progress statement in the same "batch" of data sent back to the TAIL client. With this (which is also the previous behavior that some tests ensure) you get:

```
START BATCH
1625135651092	f	1	1
1625135651092	f	1	2
1625135651093	t	\N	\N
END BATCH
```

Without this, you would get:
```
START BATCH
1625135651092	f	1	1
1625135651092	f	1	2
END BATCH
START BATCH
1625135651093	t	\N	\N
END BATCH
```

The reason (for the behaviour, not for why the client and tests want it like this) is that the input frontier is not yet "caught up" to the upper of the batch, when processing a batch. Only on the next invocation of the sink would the input frontier be at the upper of the batch, which is why we get that progress update as a separate batch when we don't emit it based on the batch upper right away.

This affects what a `FETCH ALL c` would return on the client, which is also affected by a `TIMEOUT`, if any.

Reply:
I just feels wrong to declare the upper of the batch as a new frontier if timely hasn't explicitly told us that the input hasn't advanced. Also, I don't think there are any guarantees as to how the records are broken up when you FETCH a cursor. So long as we send data with the same semantics as before we should be good. Maybe @frankmcsherry can comment more on how to handle this

Reply:
I'm very happy to drop that extra code! :+1: It's just that this is the current behaviour that tests expect.

> EDΙT: FETCH ALL c would block forever on a cursor with TAIL, no?
I think the `ALL` just means "all available", because the tests use this.

See here for reference: https://github.com/MaterializeInc/materialize/blob/main/src/materialized/tests/sql.rs#L247

Reply:
Something is guaranteeing that if we see a batch with some upper then we won't see any timestamp less than that in the future (I'm not sure what guarantees this, but we document it in the tail docs), so this should be safe. FETCH for TAIL is special and will return up to the requested number of rows as long as some are available. If no rows are available it'll return back to the user.

Reply:
It's the arrangement implementation that guarantees it. Change batches are only handed out once the frontier passes their upper. The arrangement operator will start emitting those batches but the frontier itself will not propagate to the downstream operator yet.

In a way, the upper of the batch gives us a quicker update  than waiting for the frontier update.